src/hotspot/share/gc/shenandoah/shenandoahGeneration.cpp

```diff
@@ -394,7 +394,19 @@ void ShenandoahGeneration::adjust_evacuation_budgets(ShenandoahHeap* heap, Shena
   // and promotion reserves.  Try shrinking OLD now in case that gives us a bit more runway for mutator allocations during
   // evac and update phases.
   size_t old_consumed = old_evacuated_committed + young_advance_promoted_reserve_used;
-  assert(old_available >= old_consumed, "Cannot consume more than is available");
+
+  if (old_available < old_consumed) {
+    // This can happen due to round-off errors when adding the results of truncated integer arithmetic.
+    // We've already truncated old_evacuated_committed.  Truncate young_advance_promoted_reserve_used here.
+    assert(young_advance_promoted_reserve_used <= (33 * (old_available - old_evacuated_committed)) / 32,
+           "Round-off errors should be less than 3.125%%, committed: " SIZE_FORMAT ", reserved: " SIZE_FORMAT,
+           young_advance_promoted_reserve_used, old_available - old_evacuated_committed);
+    young_advance_promoted_reserve_used = old_available - old_evacuated_committed;
+    old_consumed = old_evacuated_committed + young_advance_promoted_reserve_used;
+  }
+
+  assert(old_available >= old_consumed, "Cannot consume (" SIZE_FORMAT ") more than is available (" SIZE_FORMAT ")",
+         old_consumed, old_available);
```

Comment:
So you are just adding debugging code to catch the issue more easily? (as well as the change further below)

I don't see any change in the control flow or the state here.

Is the intention to add this change and the one below to aid debugging of a very difficult to reproduce problem?

Reply:
ah, never mind, I see that you are adjusting some values there, which I missed in a first reading of the code.

I don't understand the somewhat gnarly logic of the calculations by reading this code, but I'll assume this adjustment will then prevent whatever issue was seen in the parent ticket.

I don't understand this logic well enough to review this change, so please go ahead an land this with the review you have from William.

Reply:
```
   old_consumed = old_evacuated_committed + old_available - old_evacuated_committed
                             = old_available
```
So effectively this is saying if `old_available < old_consumed`, make `old_available = old_consumed` so the subsequent assertion is satisfied. You are asserting that this can only happen if the difference is because of round-off errors.

I do have a question about this, though. Can we avoid the round-off errors in the arithmetic entirely, by doing rounding in the "safe direction" (e.g. rounding down or up) at each step where rounding is needed, rather than relying on catching the condition and adjusting the values to re-establish the invariants?

Reply:
It would be possible to do all round-off arithmetic more conservatively, to avoid the need for this check.  The "problem" arises when the sum of individual round-off errors is different than the single round-off error for the sum of promotions.  My preference would be to fix this in a separate PR if we choose to do so.